src/react/branching.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { Realm } from "../realm.js";
+import {
+  ECMAScriptSourceFunctionValue,
+  Value,
+  UndefinedValue,
+  StringValue,
+  NumberValue,
+  BooleanValue,
+  NullValue,
+  AbstractValue,
+  ArrayValue,
+  ObjectValue,
+} from "../values/index.js";
+import { type ReactSerializerState } from "../serializer/types.js";
+import { isReactElement, addKeyToReactElement, mapOverArrayValue } from "./utils";
+import { ExpectedBailOut } from "./reconcilation.js";
+
+// Branch status is used for when Prepack returns an abstract value from a render
+// that results in a conditional path occuring. This can be problematic for reconcilation
+// as the reconciler then needs to understand if this is the start of a new branch, or if
+// it's actually deep into an existing branch. If it's a new branch, we need to apply
+// keys to the root JSX element so that it keeps it identity (because we're folding trees).
+// Furthermore, we also need to bail-out of folding class components where they have lifecycle
+// events, as we can't merge lifecycles of mutliple trees when branched reliably
+export type BranchStatusEnum = "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
+
+// Branch state is used to capture branched ReactElements so they can be analyzed and compared
+// once all branches have been processed. This allows us to add keys to the respective ReactElement
+// objects depending on various heuristics (if they have the same "type" for example)
+// A new branch state is created on a branch status of "NEW_BRANCH" and is reset to null once the branch is no
+// longer new
+export class BranchState {
+  constructor() {
+    this._branchesToValidate = [];
+  }
+  _applyBranchedLogicValue(realm: Realm, reactSerializerState: ReactSerializerState, value: Value): void {
+    if (
+      value instanceof StringValue ||
+      value instanceof NumberValue ||
+      value instanceof BooleanValue ||
+      value instanceof NullValue ||
+      value instanceof UndefinedValue
+    ) {
+      // terminal values
+    } else if (value instanceof ObjectValue && isReactElement(value)) {
+      addKeyToReactElement(realm, reactSerializerState, value);
+    } else if (value instanceof ArrayValue) {
+      mapOverArrayValue(realm, value, elementValue => {
```

Comment:
```js
<div>{[<Foo />]}<Sibling /></div>
```

Doesn't preserve state when it switches to:

```js
<div>{<Foo />}<Sibling /></div>
```

Because the nested array is in its own slot. I don't think this takes that into account.